core/src/main/java/io/jenkins/pluginhealth/scoring/probes/SCMLinkValidationProbe.java

```diff
@@ -42,20 +53,133 @@
-    public static final Pattern GH_PATTERN = Pattern.compile(GH_REGEXP);
     public static final int ORDER = UpdateCenterPluginPublicationProbe.ORDER + 100;
     public static final String KEY = "scm";
+    private static final Logger LOGGER = LoggerFactory.getLogger(SCMLinkValidationProbe.class);
+    private static final String GH_REGEXP = "https://(?<server>[^/]*)/(?<repo>jenkinsci/[^/]*)";
+    public static final Pattern GH_PATTERN = Pattern.compile(GH_REGEXP);
 
     @Override
     public ProbeResult doApply(Plugin plugin, ProbeContext context) {
         if (plugin.getScm() == null || plugin.getScm().isBlank()) {
             LOGGER.warn("{} has no SCM link", plugin.getName());
-            return ProbeResult.error(key(), "The plugin SCM link is empty");
+            return ProbeResult.error(key(), "The plugin SCM link is empty.");
+        }
+        return fromSCMLink(context, plugin.getScm(), plugin.getName());
+    }
+
+    /**
+     * Validates the SCM link, and sets {@link ProbeContext#setScmFolderPath(String)}. The value is always the path of the POM file.
+     *
+     * @param context    Refer {@link ProbeContext}.
+     * @param scm        The SCM link {@link Plugin#getScm()}.
+     * @param pluginName The name of the plugin {@link Plugin#getName()}.
+     * @return ProbeResult {@link ProbeResult}.
+     */
+    private ProbeResult fromSCMLink(ProbeContext context, String scm, String pluginName) {
+        Matcher matcher = GH_PATTERN.matcher(scm);
+        if (!matcher.find()) {
+            LOGGER.debug(String.format("%s is not respecting the SCM URL Template.", scm));
+            return ProbeResult.failure(key(), "SCM link doesn't match GitHub plugin repositories.");
+        }
+        try {
+            context.getGitHub().getRepository(matcher.group("repo"));
+            Optional<Path> pluginPathInRepository = findPluginPom(context.getScmRepository(), pluginName);
+            /*
+             * If a pom file is found, the parent folder is returned.
+             * If a pom file is not found, we assume that the plugin is a maven plugin,
+             * and return the root path. Because in a maven repository, a pom file always exists in the root.
+             */
+            Path folderPath = pluginPathInRepository
+                .flatMap(path -> Optional.ofNullable(path.getParent()))
+                .map(parent -> parent.getFileName())
+                .orElseGet(() -> {
+                    Path scmRepository = context.getScmRepository();
+                    Optional<Path> pluginPomInRootDirectory = findPluginPomInRootDirectory(context.getScmRepository(), pluginName);
+                    Path defaultFolder = pluginPomInRootDirectory.isPresent()
+                        ? scmRepository.getParent().getFileName()
+                        : Path.of("");
+                    return defaultFolder;
+                });
```

Comment:
I feel there is something you misunderstood about the file searching:

`Files.find(directory, 3, (path, $)` means that it will recursively search for the `pom.xml` file recursively, from the root directory (`directory`) up to 3 nested levels.

Meaning that if your plugin's `pom.xml` is located at the root of the repository, it will be found by `findPluginPom` directly.

There is no need for `findPluginPomInRootDirectory` as it won't catch anything else.

Your new implementation of `findPluginPom` is looking good to me. It is searching for the pom path, and if it is not found, it will return an empty optional, which is what we want.

Now for retrieving `folderPath`, what you want is:

*  If a `pom.xml` was found, the directory is its parent (cause the path ends with `/pom.xml` so we want the folder containing it)
* If no `pom.xml` was found, it is an error.

So, a simple:

```
Optional<Path> folderPath = pluginPathInRepository.map(path -> path.getParent());
```

If `folderPath` is empty, it means no `pom.xml` was found, it is an error (it will fail for Gradle plugins, but that's OK).

if `folderPath` is not empty, it will contain the path to the directory containing the `pom.xml`, which is what we want. We can use that in the `context` and return the proper result.

Reply:
One of the test cases was incorrect. And to fix that I messed up with the code.